File: src/test/java/com/soze/r2d/uibuilder/Counter.java
package com.soze.r2d.uibuilder;

import com.badlogic.gdx.scenes.scene2d.Event;
import com.soze.r2d.Component;
import com.soze.r2d.Element;
import com.soze.r2d.R;
import com.soze.r2d.UiState;

import java.util.Arrays;
import java.util.function.Function;

public class Counter extends Component {

  public Counter(UiState props) {
    super(props);
    getState().set("count", 0);
  }

  private boolean onLabelClick(Event event) {
    UiState nextState = new UiState();
    nextState.set("count", (int) getState().get("count") + 1);
    setState(nextState);
    return true;
  }

  public Element render() {
    String count = String.valueOf(getState().get("count"));
    UiState labelProps = new UiState();
    labelProps.set("text", count);
    labelProps.set("name", getProps().get("labelName"));
    labelProps.set("onClick", (Function<Event, Boolean>) this::onLabelClick);

    UiState tableProps = new UiState();
    tableProps.set("fillParent", true);
    return R.createElement("TABLE", tableProps, Arrays.asList(
      R.createElement("LABEL", labelProps)
    ));
  }
}

File: src/test/kotlin/com/soze/r2d/uibuilder/Counter.kt
package com.soze.r2d.uibuilder

import com.badlogic.gdx.scenes.scene2d.Event
import com.soze.lifegame.uibuilder.Component
import com.soze.lifegame.uibuilder.Element
import com.soze.lifegame.uibuilder.R
import com.soze.lifegame.uibuilder.UiState

class Counter(props: UiState): Component(props) {

  init {
    state["count"] = 0
  }

  private fun onLabelClick(event: Event): Boolean {
    setState("count" to (state["count"] as Int + 1))
    (props["trap"] as MutableList<String>).add("CLICKED")
    return true
  }

  override fun render(): Element {
    val count = state["count"].toString()
    val labelProps = UiState(
      "text" to "$count",
      "onClick" to this::onLabelClick,
      "name" to props["labelName"]
    )
    return R.createElement("TABLE", UiState("fillParent" to true), listOf(
      R.createElement("LABEL", labelProps)
    ))
  }
}